### DISCUSSION TITLE
Single-Stepping through a generator library

### DISCUSSION THREAD

**Discussion by icebeing** [halide/Halide#7996] [Q&A] [answered]
Hi all -

   I've managed to build my algorithm's generator & runner *.cpp sources with a "debug" configuration:  using -O0, -g3 clang++ flags.
I am able to single-step through the runner's main(), with lldb

However, when I step into the generator's function call, all I see is native assembly.  
My guess is the myalgo_generator executable strips symbols from my libmyalgo.a library.  Is there an option in the generator exec to turn on source-stepping within the library?

Thanks, Charles.

**Comment by abadams** [comment, marked as answer]
No, Halide isn't an imperative language, so there's no good correspondence between the lines of code that define your Funcs and the steps taken at runtime. Something that might achieve your aims is the -trace_all Halide target flag, which will print out everything Halide does. If you're getting a crash, the -debug target flag is also useful for verifying your input and output buffers are what you expect.

Another option is to compile Halide to c++ with -e c_source, compile and link against that, and single-step through that code instead. Compiling Halide to C++ typically produces code with worse performance than compiling Halide directly to an object file, so I'd only use it for debugging.

**Reply by icebeing** [reply]
OK, I will investigate those points. Thanks for clarifying.
